### PR TITLE
IJSDK-479A

### PR DESCRIPTION
**Pull request by JohnHake**
This is the code portion of IJSDK-479 Convert Simple Language Code Sample to Gradle. Part B will be a review of the updated tutorials for custom language support and testing a custom language.

### CODE REVIEW

File: code_samples/simple_language/build.gradle [hunk 1]
```diff
@@ -0,0 +1,37 @@
+plugins {
+  id 'java'
+  id 'org.jetbrains.intellij' version '0.4.15'
```
Comment (YannCebron) [inline code_samples/simple_language/build.gradle]: 0.4.16

File: code_samples/simple_language/build.gradle [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+sourceCompatibility = 1.8
+
+repositories {
+  mavenCentral()
+}
+
+test {
+  // Set idea.home.path to the absolute path to the intellij-community source
+  // on your local machine.
+  systemProperty "idea.home.path", "/Users/jhake/Documents/source/comm"
```
Comment (YannCebron) [inline code_samples/simple_language/build.gradle]: let's use something that clearly indicates this is a placeholder text

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleAnnotator.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+
+public class SimpleAnnotator implements Annotator {
+  // Define strings for the Simple language prefix - used for annotations, line markers, etc.
+  public static final String SIMPLE_PREFIX_STR = "simple";
+  public static final String SIMPLE_SEPARATOR_STR = ":";
+  
+  @Override
+  public void annotate(@NotNull final PsiElement element, @NotNull AnnotationHolder holder) {
+    // Ensure the Psi element is an expression
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleAnnotator.java]: PSI Element

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleAnnotator.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    if ( !( element instanceof PsiLiteralExpression ) ) return;
+    
+    // Ensure the Psi element contains a string that starts with the key and separator
+    PsiLiteralExpression literalExpression = (PsiLiteralExpression) element;
+    String value = literalExpression.getValue() instanceof String ? (String) literalExpression.getValue() : null;
+    if ( ( value == null ) || !value.startsWith( SIMPLE_PREFIX_STR + SIMPLE_SEPARATOR_STR ) ) return;
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleAnnotator.java]: where do all these whitespaces around `(` and `)` come from? is this really the default formatting rules?

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleAnnotator.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    // "simple:key"
+    //  01234567890
+    TextRange prefixRange = TextRange.from( element.getTextRange().getStartOffset(), SIMPLE_PREFIX_STR.length() + 1 );
+    TextRange separatorRange = TextRange.from( prefixRange.getEndOffset(), SIMPLE_SEPARATOR_STR.length() );
+    TextRange keyRange = new TextRange( separatorRange.getEndOffset(), element.getTextRange().getEndOffset() - 1 );
+    
+    // Get the list of properties from the Project
+    String possibleProperties = value.substring( SIMPLE_PREFIX_STR.length() + SIMPLE_SEPARATOR_STR.length() );
+    Project project = element.getProject();
+    List< SimpleProperty > properties = SimpleUtil.findProperties( project, possibleProperties );
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleAnnotator.java]: weird whitespace around generics

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleCompletionContributor.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import com.intellij.codeInsight.completion.*;
+import com.intellij.codeInsight.lookup.LookupElementBuilder;
+import com.intellij.patterns.PlatformPatterns;
+import com.intellij.util.ProcessingContext;
+import com.intellij.sdk.language.psi.SimpleTypes;
+import org.jetbrains.annotations.NotNull;
+
+public class SimpleCompletionContributor extends CompletionContributor {
+  public SimpleCompletionContributor() {
+    // Register the completion providers
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleCompletionContributor.java]: let's remove all these line comments

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  @Override
+  public FoldingDescriptor[] buildFoldRegions(@NotNull PsiElement root, @NotNull Document document, boolean quick) {
+    // Initialize the group of folding regions that will expand/collapse together.
+    FoldingGroup group = FoldingGroup.newGroup(SIMPLE_PREFIX_STR);
+    // Initialize the list of folding regions
+    List< FoldingDescriptor > descriptors = new ArrayList< FoldingDescriptor >();
+    // Get a collection of the literal expressions in the document below root
+    Collection< PsiLiteralExpression > literalExpressions =
+          PsiTreeUtil.findChildrenOfType(root, PsiLiteralExpression.class);
+    // Evaluate the collection, using only Simple language literalExpressions
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java]: what does "using only Simple language literalExpressions" mean?

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    for ( final PsiLiteralExpression literalExpression : literalExpressions ) {
+      String value = literalExpression.getValue() instanceof String ? (String) literalExpression.getValue() : null;
+      if ( value != null && value.startsWith(SIMPLE_PREFIX_STR + SIMPLE_SEPARATOR_STR) ) {
+        Project project = literalExpression.getProject();
+        String key = value.substring(SIMPLE_PREFIX_STR.length() + SIMPLE_SEPARATOR_STR.length());
+        // Get a list of properties for the project
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java]: "find all properties for given key in project"

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    return descriptors.toArray(new FoldingDescriptor[descriptors.size()]);
+  }
+  
+  /**
+   * Gets the Simple Language 'value' string corresponding to the 'key'
+   * @param node  Node corresponding to PsiLiteralExpression containing a string in the format
+   *              SIMPLE_PREFIX_STR + SIMPLE_SEPARATOR_STR + Key, where Key is
+   *              defined by the Simple language file.
+   * @return String corresponding to the "website" key.
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java]: remove `@return` comment completely

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleRefactoringSupportProvider.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import org.jetbrains.annotations.*;
+
+public class SimpleRefactoringSupportProvider extends RefactoringSupportProvider {
+  @Override
+  public boolean isMemberInplaceRenameAvailable(@NotNull PsiElement elementToRename, @Nullable PsiElement context) {
+    return (elementToRename instanceof SimpleProperty);
+  }
+}
+
+/*
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleRefactoringSupportProvider.java]: ??

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleReferenceContributor.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import com.intellij.psi.*;
+import com.intellij.util.ProcessingContext;
+import org.jetbrains.annotations.NotNull;
+
+import static com.intellij.sdk.language.SimpleAnnotator.*;
+
+public class SimpleReferenceContributor extends PsiReferenceContributor {
+  @Override
+  public void registerReferenceProviders(@NotNull PsiReferenceRegistrar registrar) {
+    registrar.registerReferenceProvider( PlatformPatterns.psiElement( PsiLiteralExpression.class ),
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleReferenceContributor.java]: please reformat to have less indentation in `PsiReferenceProvider` implementation

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleStructureViewElement.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  @Override
+  public TreeElement[] getChildren() {
+    if (element instanceof SimpleFile) {
+      SimpleProperty[] properties = PsiTreeUtil.getChildrenOfType(element, SimpleProperty.class);
+      List<TreeElement> treeElements = new ArrayList<TreeElement>(properties.length);
+      for (SimpleProperty property : properties) {
+        treeElements.add(new SimpleStructureViewElement((SimplePropertyImpl) property));
+      }
+      return treeElements.toArray(new TreeElement[treeElements.size()]);
+    } else {
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleStructureViewElement.java]: superfluous `else`

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleUtil.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import com.intellij.psi.search.*;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.intellij.util.indexing.FileBasedIndex;
+import com.intellij.sdk.language.psi.*;
+
+import java.util.*;
+
+public class SimpleUtil {
+  
+  // Searches the entire project for Simple language files with instances of the Simple property
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleUtil.java]: ... "with given key"

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/psi/SimpleFile.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  public FileType getFileType() {
+    return SimpleFileType.INSTANCE;
+  }
+  
+  @Override
+  public String toString() {
+    return "Simple File";
+  }
+  
+  @Override
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/java/com/intellij/sdk/language/psi/SimpleFile.java]: remove `getIcon()` method completely

File: code_samples/simple_language/src/main/resources/META-INF/plugin.xml [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+  <id>com.intellij.sdk.simple_language</id>
+
+  <!-- Text to display as name on Preferences/Settings | Plugin page  -->
+  <name>SDK: Simple Language Sample Project</name>
+
+  <!-- The version of this plugin -->
+  <version>2.0.0</version>
+
+  <!-- Compatible with the following versions of IntelliJ Platform
+        At least 2019.2 is required. Otherwise, a FileTypeFactor must be registered as an extension.
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/resources/META-INF/plugin.xml]: typo `FileTypeFactor`Y

File: code_samples/simple_language/src/main/resources/META-INF/plugin.xml [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    ]]>
+  </change-notes>
+
+  <!-- Text to display as company information on Preferences/Settings | Plugin page -->
+  <vendor url="https://plugins.jetbrains.com">IntelliJ Platform SDK</vendor>
+
+  <extensions defaultExtensionNs="com.intellij">
+    <fileType name="Simple file" implementationClass="com.intellij.sdk.language.SimpleFileType" fieldName="INSTANCE"
+            language="Simple" extensions="simple"/>
+    <!-- Only required for versions of the IntelliJ Platform prior to v2019.2.
```
Comment (YannCebron) [inline code_samples/simple_language/src/main/resources/META-INF/plugin.xml]: comment that it has to be used _instead_ of `fileType` EP

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleCompletionContributor.java (hunk 6 above)
Comment (JohnHake) [reply code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleCompletionContributor.java]: I removed the comments...but let's keep in mind this _isn't_ production code.

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java (hunk 7 above)
Comment (JohnHake) [reply code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleFoldingBuilder.java]: You're right, it's redundant.

File: code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleRefactoringSupportProvider.java (hunk 10 above)
Comment (JohnHake) [reply code_samples/simple_language/src/main/java/com/intellij/sdk/language/SimpleRefactoringSupportProvider.java]: One of the errors I occasionally see regarding Kotlin. I have removed it.

**Pull request merged**